### ISSUE TITLE
duckling component returns numbers as time

### ISSUE THREAD

**Issue by ecomaven** [golastmile/rasa_nlu#310]
**rasa NLU version** (e.g. `0.7.3`): 0.8.0all

**Used backend / pipeline** (`mitie`, `spacy_sklearn`, ...): [ "nlp_mitie", "tokenizer_mitie", "ner_mitie", "ner_duckling", "ner_synonyms", "intent_featurizer_mitie", "intent_classifier_sklearn" ]

**Operating system** (windows, osx, ...): windows/vagrant/virtualenv

**Issue**: 
When using the ner_duckling component setting `duckling_processing_mode` as either append or replace, it returns numbers in any input ( e.g. 'send 5000') as time -> `duckling: time` and `"value": "5000-01-01T00:00:00.000Z"`.

From [Duckling documentation](https://duckling.wit.ai/#getting-started) it is possible to specify the `core` modules to be used. Seems only the `:time` was set. 

There is no way to specify the duckling `core` modules to use in rasa configuration or am I missing something? 

**Content of configuration file** (if used & relevant):
```
{
  "pipeline": [ "nlp_mitie", "tokenizer_mitie", "ner_mitie", "ner_duckling", "ner_synonyms", "intent_featurizer_mitie", "intent_classifier_sklearn" ],
  "mitie_file": "./data/total_word_feature_extractor.dat",
  "num_threads": 4,
  "path" : "./models_mitie_sklearn",
  "data" : "/devFolder/rasaData/training_data.json"
}
```

**Comment by tmbo** [comment]
@twerkmeister alright, let me try to explain this. It is related to two other issues that all can be solved with one solution for the underlying problem (#319, #300):
The return values from the `process` function of each component are glued together, in a very simple way by taking the dictionary values returned from the current component and overriding any values that already exist with the same key. Hence, the default behaviour is to replace existing keys, if a component now wants to append new values it needs to do that internally and return the result. Example:
Lets assume the current processing context is 
```json
{
   "text": "How is the weather in Berlin tomorrow",
   "intent": { ... }, 
   "entities": [{"entity": "city": "value": "Berlin"}]
}
```
meaning, this is the output previous components created before our current component. If our current component now wants to append to the entities array, it needs to do that in its process method on its own. If the component would just return 
```json
{
   "entities": [{"entity": "date": "value": "tomorrow"}]
}
the resulting context (and later the output) will be
```json
{
   "text": "How is the weather in Berlin tomorrow",
   "intent": { ... }, 
   "entities": [{"entity": "date": "value": "tomorrow"}]
}
```
The previous entity results are replaced with the output from our component (which is most likely not what we want for entites, assuming we use different NER components to combine their results). 

There are different merge strategies we need:
- merging for the `"intent"` value: the one with the highest score should win (would allow us to create a keyword matching component that selects intent based on keywords, but if no keword is present one of the existing classifiers is used )
- apending to `"entities"`
- replacing single results in `"entities"` (e.g. the synonym mapper needs to replace a result in the entities array if it found a synonym and needs to update the value of this result)

The issue with the duckling component is due to the above and another quirk of the duckling implementation: If the sentence contains a an entity duckling recognizes, e.g. "in 2 weeks", duckling will return multiple entities with different token spans for its different supported types (e.g. for "in 2 weeks" it might return `[{"entity": "date", "value": "2 weeks"}, {"entity": "number", "value": "2"}, {"enity": "length", "value": "2"}, ...]`).

**Comment by twerkmeister** [comment]
@tmbo thanks for sharing this!

**Comment by twerkmeister** [comment]
Actually, the merging seems to make sense, except for the case that no prior entity list is present #300 . The problem here is, that we are only taking the first dimension of the match, which is always the time dimension. Unfortunately duckling does not provide a confidence score for its matches, so we can't just take the highest confidence match. I propose that we let the user define which dimensions are interesting to them, and we forward all matches from these dimensions. If multiple defined dimensions match, the user has to determine which makes most sense for the recognized intent.

**Comment by twerkmeister** [comment]
@tmbo I think in that sense the `replace` mode for duckling doesn't make sense. It only makes sense if the user is interested in a single duckling dimension. Otherwise we might have multiple dimensions and we don't know which one to pick. So we would replace one entity with multiple options. That, in my view obliterates the use case for replacement in the first place: that for a given entity there is only one result and the user can go ahead with it. 

What would you think about getting rid of replacement and actually always appending + adding which component extracted the given entity. Based on that users could add their own components down the pipeline that merge the results, based on their use case.
For example they could:
* always taking the most specific entity (largest extracted chunk)
* Use inter-extractor agreement to determine a entity type
* possibly more, as such a component would have much more information available
Also this solves the problem of tweaking the order of extractors, a process which is probably gonna be categorized as black magic :D

**Comment by tmbo** [comment]
@twerkmeister right, I think we are converging on this one. This is what I mean with externalization of the merging logic ;) 

I think there are use cases for the replace mode but I also spotted the issue you mentioned which is the one a user experienced as "duckling always responds with time". Let's remove it for now and keep an eye on users and our needs with this.

The merging sounds great. Ideally we need this for entities and intent at the moment (where the merging on intent should probably done based on the confidence). I am not sure but we might need to adapt the format here before we release 0.8 
Currently:
```json
{
  "intent": {
        "confidence": 0.5192305466357352,
        "name": "restaurant_search"
    },
    "intent_ranking": [
        {
            "confidence": 0.5192305466357352,
            "name": "restaurant_search"
        },
        {
            "confidence": 0.2066287604378098,
            "name": "affirm"
        }
    ]
}
```
Changed:
```json
    "intent": {
        "confidence": 0.5192305466357352,
        "name": "restaurant_search",
        "ranking": [
            {
                "confidence": 0.5192305466357352,
                "name": "restaurant_search"
            },
            {
                "confidence": 0.2066287604378098,
                "name": "affirm"
            }
        ]
    }
```
Since we need to use the intent ranking of the component that was most confident in the final intent it selected... But this is something you need to decide when implementing it.

**Comment by tmbo** [comment]
I think this should be fixed with #324 right, @twerkmeister ?
